mp/src/game/shared/momentum/run/mom_replay_factory.cpp

```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+                                                                 MomLocalReplayCallback_t callback,
+                                                                 MomReplayAsyncHandle* outHandle)
+{
+    auto *internalRequest = new MomReplayLoadRequestInternal;
+    if (outHandle)
+        *outHandle = internalRequest->handle;
+    requests.AddToTail(internalRequest);
+    internalRequest->callback = callback;
+
+    size_t count = filenames.Size();
+
+    auto fsControls = new FSAsyncControl_t[count];
+    auto fsRequests = new FileAsyncRequest_t[count]; // they SHOULD get deleted by filesystem
+
+    internalRequest->arrControls = fsControls;
+    internalRequest->controls_count = count;
+    internalRequest->controls_pending = count;
+
+    FOR_EACH_VEC (filenames, i)
+    {
+        fsRequests[i].pszFilename = filenames[i];
+        fsRequests[i].pfnCallback = FilesystemCallbackForward;
+        fsRequests[i].pszPathID = pathID;
+        fsRequests[i].pContext = internalRequest->handle;
+    }
+
+    if (filesystem->AsyncReadMultiple(fsRequests, count, fsControls) != FSASYNC_OK)
+    {
+        Warning("Bad async replay load request!");
+        AsyncCleanup(internalRequest->handle);
+        return LocalTimesStatus_t::LOCALREPLAY_BADREQUEST;
+    }
+
+    return LocalTimesStatus_t::LOCALREPLAY_INPROGRESS;
+
+}
+void CMomReplayFactory::FilesystemCallbackForward(const FileAsyncRequest_t &request, int nBytesRead,
+                                                  FSAsyncStatus_t err)
+{
+    g_ReplayFactory.FilesystemCallback(request, nBytesRead, err);
+}
+void CMomReplayFactory::FilesystemCallback(const FileAsyncRequest_t &request, int nBytesRead, FSAsyncStatus_t err)
+{
+    auto handle = reinterpret_cast<MomReplayAsyncHandle>(request.pContext);
+    auto intReq = FindRequestByHandle(handle);
+
+    Assert(intReq);
+
+
+
+    if (err != FSASYNC_OK)
+    {
+        // This shouldn't happen, we just cancel the whole operation
+        free(request.pData);
+        AsyncCleanup(handle);
+    }
+
+    // Build Buffer and create replay
+
+    CUtlBuffer reader(request.pData, nBytesRead, CUtlBuffer::READ_ONLY);
+    reader.SeekPut(CUtlBuffer::SEEK_TAIL, 0); // The hash function will break otherwise!
+
+    auto replay = LoadReplayFromBuffer(reader, false);
+    intReq->replayMutex.Lock();
+    if (replay)
+    {
+        intReq->replays.AddToTail(replay);
+    }
+
+    --(intReq->controls_pending);
+
+
+    intReq->replayMutex.Unlock();
+
+    Assert(intReq->controls_pending >= 0);
+
+    if (intReq->controls_pending == 0)
+    {
+        // We are done here and ready to call back
+        intReq->callback();
+    }
+}
+LocalTimesStatus_t CMomReplayFactory::AsyncGetReplays(MomReplayAsyncHandle handle,
+                                                             CUtlVector<CMomReplayBase*> &vecToReturn,
+                                                             bool bCancelRemaining)
```

Comment:
I feel like this is doing a bit too much. Canceling the remaining upon trying to read this just doesn't make sense to me as a use case. You either want them all when they're ready, or you cancel the requests (as you shutdown map / the game).

I mean maybe you can make an overload to the Cancel that's like `CancelAndFillPartial` but I am still trying to understand when this would be the desired case.